src/vm-vcpu-ref/src/x86_64/gdt.rs

```diff
@@ -1,114 +1,254 @@
+        kvm_segment {
+            base: self.base(),
+            limit: self.limit(),
+            selector: u16::from(table_index * 8),
+            type_: self.segment_type(),
+            present: self.p(),
+            dpl: self.dpl(),
+            db: self.db(),
+            s: self.s(),
+            l: self.l(),
+            g: self.g(),
+            avl: self.avl(),
+            padding: 0,
+            unusable: match self.p() {
+                0 => 1,
+                _ => 0,
+            },
+        }
+    }
 }
 
-pub fn get_s(entry: u64) -> u8 {
-    ((entry & 0x0000_1000_0000_0000) >> 44) as u8
+#[derive(Clone, Debug)]
+/// The `Gdt` is a wrapper for creating and managing operations on the
+/// Global Descriptor Table (GDT). The GDT is a data structure used by
+/// Intel x86-family processors to define the characteristics of the
+/// various memory areas used during program execution (like data, code, TSS).
+///
+/// The `Gdt` provides a default implementation that can be used for setting
+/// up a vCPU for booting. The default implementation contains all 4 segment
+/// descriptors corresponding to null, code, data, and TSS. The default can be
+/// extended by using `push`. When the default is not matching the product
+/// requirements, the GDT can also be created from scratch.
+pub struct Gdt(Vec<SegmentDescriptor>);
+
+impl Gdt {
+    /// Create an empty `GDT`.
+    ///
+    /// # Example - Creating a GDT from scratch
+    ///
+    /// ```rust
+    /// use vm_vcpu_ref::x86_64::gdt::{Gdt, SegmentDescriptor};
+    ///
+    /// let mut gdt = Gdt::new();
+    /// // Create the GDT as in the GDT Tutorial: https://wiki.osdev.org/GDT_Tutorial
+    /// gdt.push(SegmentDescriptor::from(0, 0, 0));
+    /// gdt.push(SegmentDescriptor::from(0x9A, 0, 0xffffffff));
+    /// gdt.push(SegmentDescriptor::from(0x92, 0, 0xffffffff));
+    /// gdt.push(SegmentDescriptor::from(0x89, 0, 0xffffffff));
+    /// ```
+    pub fn new() -> Gdt {
+        Gdt(vec![])
+    }
+
+    /// Writes the GDT (Global Descriptor Table) into Guest Memory.
+    ///
+    /// # Example - Creating the GDT for Linux
+    ///
+    /// ```rust
+    /// use vm_memory::{GuestAddress, GuestMemoryMmap};
+    /// use vm_vcpu_ref::x86_64::gdt::Gdt;
+    ///
+    /// let guest_memory: GuestMemoryMmap =
+    ///     GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 1024 << 20)]).unwrap();
+    /// let gdt_table = Gdt::default();
+    /// gdt_table.write_to_mem(&guest_memory).unwrap();
+    /// ```
+    pub fn write_to_mem<Memory: GuestMemory>(
+        &self,
+        mem: &Memory,
+    ) -> std::result::Result<(), vm_memory::GuestMemoryError> {
+        let boot_gdt_addr = GuestAddress(BOOT_GDT_OFFSET);
+        for (index, entry) in self.iter().enumerate() {
+            let addr = mem
+                .checked_offset(boot_gdt_addr, index * mem::size_of::<u64>())
+                .ok_or(vm_memory::GuestMemoryError::InvalidGuestAddress(
```

Comment:
Somehow `clippy` always complains that I should use `.ok_or_else(|| {...})` with a closure instead of creating an error value regardless of the control flow :)

Reply:
`ok_or` is eagerly evaluated, while `ok_or_else` is lazy evaluated. In this particular case we're not calling any functions, so I think we should be okay with using `ok_or`.